### ISSUE TITLE
Sharding whishlist: Add `OpShardingSharding.reshape`, consider `sharding.insert_axis`

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#13666]
I would like to be able to reshape an `OpShardingSharding` object. 
This would be useful to implement sharding-aware functions that use the sharding of the input `jax.Array` in their logic.

Some shadings like `PositionalSharding` support the reshape, but as soon as you pass a `jax.Array` with `PositionalSharding` to a `jax.jit` you get back an array with `OpShardingSharding` which cannot be reshaped anymore. See the MWE:

```python
import os
os.environ["XLA_FLAGS"]="--xla_force_host_platform_device_count=4"
import jax
import jax.numpy as jnp

devices = jax.devices()
sharding = jax.sharding.PositionalSharding(devices)

x = jax.device_put(jnp.ones((16,16)), sharding.reshape(1,-1))
y = jax.jit(lambda x,c: x+c)(x,1.3)

In [8]: print(x.sharding)
PositionalSharding([[{CPU 0} {CPU 1} {CPU 2} {CPU 3}]])

In [9]: print(y.sharding)
OpShardingSharding({devices=[1,4]0,1,2,3})
```

In addition, it would also be useful (and less error prone) to be able to directly add a new axis to a sharding instead of having to build the shape tuple by hand, for example by being able to use `sharding.insert_axis`.

**Comment by yashk2810** [comment]
JAX shouldn't output GSPMDSharding anymore. If you ever see that, please open another issue. Until then, I am going to close this since you will never see GSPMDSharding as an output.
